### PR TITLE
[10.0][MIG] stock_cycle_count

### PR DESCRIPTION
**Pull request by LoisRForgeFlow**
Migration from v9 to v10.

Depends on:
* [x] : https://github.com/OCA/stock-logistics-warehouse/pull/319
* [x] : https://github.com/OCA/stock-logistics-warehouse/pull/318
* [x] : https://github.com/OCA/stock-logistics-warehouse/pull/330

### CODE REVIEW

**Comment by LoisRForgeFlow** [comment]
@alan196 @rousseldenis you might be interested in this one

File: stock_cycle_count/models/stock_inventory.py
Comment (MiquelRForgeFlow) [inline stock_cycle_count/models/stock_inventory.py]: 100.0 → PERCENT

Comment (LoisRForgeFlow) [reply stock_cycle_count/models/stock_inventory.py]: Sure, thanks

File: stock_cycle_count/models/stock_cycle_count.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count.py]: Following https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#models
the method name should be _compute...

File: stock_cycle_count/models/stock_cycle_count.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count.py]: @lreficent This one should have been decorated multi with ensure_one. Inconsistency between decorator and self value (not void recordset)

File: stock_cycle_count/models/stock_cycle_count_rule.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count_rule.py]: Deprecated

File: stock_cycle_count/models/stock_cycle_count_rule.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count_rule.py]: Deprecated

File: stock_cycle_count/models/stock_cycle_count_rule.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count_rule.py]: @lreficent self can be a multi recordset

File: stock_cycle_count/models/stock_cycle_count_rule.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count_rule.py]: string can be omitted (and others here under)

File: stock_cycle_count/models/stock_cycle_count_rule.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count_rule.py]: __compute_...

File: stock_cycle_count/models/stock_cycle_count_rule.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count_rule.py:170]: What can be the type of this Exception? Maybe this can be more specific

File: stock_cycle_count/models/stock_cycle_count_rule.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count_rule.py]: Same inconsistency with self and decorator

File: stock_cycle_count/models/stock_inventory.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_inventory.py]: Maybe this can be written like `self.filtered(lambda i: i.cycle_count_id and i.state =='done').mapped('cycle_count_id').state = 'done'`

File: stock_cycle_count/models/stock_inventory.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_inventory.py:56]: Maybe this should be done in a constrains ?

File: stock_cycle_count/models/stock_location.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_location.py]: Deprecated

File: stock_cycle_count/models/stock_location.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_location.py]: This is default

File: stock_cycle_count/models/stock_location.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_location.py]: self / decorator

File: stock_cycle_count/models/stock_location.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_location.py]: Deprecated

File: stock_cycle_count/models/stock_location.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_location.py]: @api.multi, self.ensure_one()

File: stock_cycle_count/models/stock_warehouse.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_warehouse.py]: Deprecated

File: stock_cycle_count/models/stock_warehouse.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_warehouse.py]: Deprecated

File: stock_cycle_count/models/stock_warehouse.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_warehouse.py]: To have more clearer code, could you use self.env[model].search() here ?

File: stock_cycle_count/models/stock_warehouse.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_warehouse.py]: I think this works. But quite weird. I would have done `('warehouse_ids', 'in', [self.id])`

**Review by rousseldenis — CHANGES_REQUESTED** [review]
LGTM but there are IMHO lot of api issues

@lreficent Can you fix them ?

**Comment by LoisRForgeFlow** [comment]
* [x] : Open improvements in v9: https://github.com/OCA/stock-logistics-warehouse/pull/346 EDIT: added

@rousseldenis yes, I have to clean all those api usages. I will do after the above improvements for v9 are merged

File: stock_cycle_count/models/stock_inventory.py
Comment (LoisRForgeFlow) [reply stock_cycle_count/models/stock_inventory.py:56]: You cannot modify the inventory at all in this case so I do not see a specific constrain here.

File: stock_cycle_count/models/stock_cycle_count_rule.py
Comment (LoisRForgeFlow) [reply stock_cycle_count/models/stock_cycle_count_rule.py:170]: Well, I could put a `ZeroDivisionError` but I prefer to leave it open

**Comment by LoisRForgeFlow** [comment]
@rousseldenis Finally I attended your review, Thanks!

**Comment by gurneyalex** [comment]
I triggered a runbot rebuild

**Review by gurneyalex — APPROVED** [review]

File: stock_cycle_count/models/stock_cycle_count.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count.py]: Please follow OCA conventions and set fields definitions above.

File: stock_cycle_count/models/stock_cycle_count.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count.py]:
```python
@api.depends('stock_adjustment_ids')
```

File: stock_cycle_count/models/stock_cycle_count_rule.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count_rule.py]: Please follow OCA convention for onchange function name

File: stock_cycle_count/models/stock_cycle_count_rule.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count_rule.py]: Same as above.

File: stock_cycle_count/models/stock_cycle_count_rule.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count_rule.py]: Maybe the search can be directly order by date desc and limit=1 to fetch less records.

File: stock_cycle_count/models/stock_location.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_location.py:26]: Maybe it could be interesting to have an inventory history field on locations ?

File: stock_cycle_count/models/stock_move.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_move.py]: check_zero_confirmation() function is a real multi, maybe it could be called on all locations ?

File: stock_cycle_count/models/stock_warehouse.py
Comment (rousseldenis) [inline stock_cycle_count/models/stock_warehouse.py]: IMHO, maybe this function could be refactorized as I see some searched records that can become fields(that could help reading - and some could be called on other code parts ?).

And the values set on the create at the end could be retrieved through a _prepare hook function.

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Some comments

**Comment by gurneyalex** [comment]
runbot status unrelated. @lreficent  if you rebase it should get green.

File: stock_cycle_count/models/stock_cycle_count_rule.py
Comment (LoisRForgeFlow) [reply stock_cycle_count/models/stock_cycle_count_rule.py]: it's both a compute method and an onchange method. https://github.com/Eficent/stock-logistics-warehouse/blob/b9b4c2283fdffebbcdb508e49e07a5e512d70cb6/stock_cycle_count/models/stock_cycle_count_rule.py#L100

File: stock_cycle_count/models/stock_cycle_count.py
Comment (LoisRForgeFlow) [reply stock_cycle_count/models/stock_cycle_count.py]: done

File: stock_cycle_count/models/stock_cycle_count.py
Comment (LoisRForgeFlow) [reply stock_cycle_count/models/stock_cycle_count.py]: done

File: stock_cycle_count/models/stock_cycle_count_rule.py
Comment (LoisRForgeFlow) [reply stock_cycle_count/models/stock_cycle_count_rule.py]: fixed

File: stock_cycle_count/models/stock_cycle_count_rule.py
Comment (rousseldenis) [reply stock_cycle_count/models/stock_cycle_count_rule.py]: If you put
```python
@api.depends('rule_type')
```
The change on rule_type value will trigger the change on rule_description on the interface.

File: stock_cycle_count/models/stock_cycle_count_rule.py
Comment (LoisRForgeFlow) [reply stock_cycle_count/models/stock_cycle_count_rule.py]: right, done.

File: stock_cycle_count/models/stock_location.py
Comment (LoisRForgeFlow) [reply stock_cycle_count/models/stock_location.py:26]: what kind of field, could you please explain a little bit more?

File: stock_cycle_count/models/stock_warehouse.py
Comment (LoisRForgeFlow) [reply stock_cycle_count/models/stock_warehouse.py]: agree, but I can't do that at the moment, I would look into it when I have some spare time :)

File: stock_cycle_count/models/stock_move.py
Comment (LoisRForgeFlow) [reply stock_cycle_count/models/stock_move.py]: done

File: stock_cycle_count/models/stock_location.py
Comment (rousseldenis) [reply stock_cycle_count/models/stock_location.py:26]: An 'inventory_history_ids' that represent inventories done on a location.

File: stock_cycle_count/models/stock_cycle_count_rule.py
Comment (LoisRForgeFlow) [reply stock_cycle_count/models/stock_cycle_count_rule.py]: done

**Review by rousseldenis — APPROVED** [review]
Code review.

Should add Roadmap for further version to improve the action_compute_cycle_count_rules in stock.warehouse

**Comment by LoisRForgeFlow** [comment]
@rousseldenis would you do the honors of merging this? :wink:

**Comment by rousseldenis** [comment]
@lreficent Wait for all green checks, but yes of course!

**Pull request merged**
